### PR TITLE
AdHocVariables: Fix an issue where adhoc filter values stored as integer was causing Grafana to crash

### LINKED ISSUE #31335: Following upgrade from 7.3.7 to 7.4.2 - adhoc filter values stored as integer cause e.replace error
**What happened**:
Upgrade from 7.3.7 to 7.4.2, dashboard will not load properly, instead receive e.replace error
**What you expected to happen**:
Dashboard to load as under 7.3.7

**How to reproduce it (as minimally and precisely as possible)**:
Create a dashboard in 7.3.7.  Add Elasticsearch Data Source.  Add a Table Panel to view the data.  Add an adhoc filter, using a key/ value where the value is a number.  Save the dashboard, and select the option "Save current variable values as dashboard default". You can now go into the dashboard settings and view the JSON to see similar to the snippet below, the value will be stored as an integers (unquoted number). Upgrade to 7.4.2.  Attempt to open dashboard - e.replace error :-)

**Anything else we need to know?**:
I already spent a few hours diff-ing the cause.

In 7.3.7, here is the JSON snippet of the Dashboard.  note the "filters" key, and the value, stored as a JSON Integer (4728):

```
{
          "datasource": "$datasource",
          "error": null,
          "filters": [
            {
              "condition": "",
              "key": "winlog.event_id",
              "operator": "=",
              "value": 4728
            }
          ],
          "hide": 0,
          "label": null,
          "name": "Filters",
          "skipUrlSync": false,
          "type": "adhoc"
        }
```

If upgraded to 7.4.2, the result is the "e.replace" error message.  If you create a new dashboard in 7.4.2, I found the value is stored as a string instead:

```
{
        "datasource": "$datasource",
        "description": null,
        "error": null,
        "filters": [
          {
            "condition": "",
            "key": "winlog.event_id",
            "operator": "=",
            "value": "4728"
          }
        ],
        "hide": 0,
        "label": null,
        "name": "Filters",
        "skipUrlSync": false,
        "type": "adhoc"
      }
```

Probably related to code changes with the adhoc filters (I see some related to regex).

**Environment**:
- Grafana version: 7.4.2 (from 7.3.7, which was working)
- Data source type & version: Elasticsearch 7.9.2
- OS Grafana is installed on: Centos 7
- User OS & Browser: MacOS - Safari and Chrome
- Grafana plugins: Elasticsearch Datasource, Table Plugin
- Others:

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
Looked into storing everything as strings but that could probably have other effects so I ended up just fixing the issue and adding some tests.

**Which issue(s) this PR fixes**:
Fixes #31335

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/variables/adhoc/urlParser.ts [hunk 1]
```diff
@@ -17,11 +17,19 @@ export const toFilters = (value: UrlQueryValue): AdHocVariableFilter[] => {
 };
 
 function escapeDelimiter(value: string | undefined): string {
-  return value?.replace(/\|/g, '__gfp__') ?? '';
+  if (value === null || value === undefined) {
+    return '';
+  }
+
+  return /\|/g[Symbol.replace](value, '__gfp__');
```
Comment (dprokop) [inline public/app/features/variables/adhoc/urlParser.ts:24]: @hugohaggmark interesting, can you say few words how does this help?

**Review by dprokop — APPROVED** [review]

File: public/app/features/variables/adhoc/urlParser.ts (hunk 1 above)
Comment (hugohaggmark) [reply public/app/features/variables/adhoc/urlParser.ts:24]: That method handles numbers, strings, boolean etc but `null` and `undefined` are just converted to `"null"` and `"undefined"` so I still need the if statement unfortunately.

Comment (dprokop) [reply public/app/features/variables/adhoc/urlParser.ts:24]: OK, I see, thanks for the explanation <3

Comment (hugohaggmark) [reply public/app/features/variables/adhoc/urlParser.ts:24]: It's like string.replace method but on a regex instance. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/@@replace#direct_call

**Comment by davewat** [comment]
Tested against nightly build - resolves issue #31335.  Thank you!

**Comment by hugohaggmark** [comment]
> Tested against nightly build - resolves issue #31335. Thank you!

Wow, thank you for taking the time to test that @davewat 🎉

**Pull request merged**